File: benchmarks/opt2-ab/window.c
// Read only: report the first ordinary window owned by one test process.
#include <CoreGraphics/CoreGraphics.h>
#include <CoreFoundation/CoreFoundation.h>
#include <errno.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>
int main(int argc,char **argv) {
    if(argc!=2)return 2;
    char *end;errno=0;long pid=strtol(argv[1],&end,10);
    if(errno||*end||pid<1||pid>INT_MAX)return 2;
    for(int attempt=0;attempt<600;attempt++) {
        CFArrayRef a=CGWindowListCopyWindowInfo(kCGWindowListOptionAll,kCGNullWindowID);
        if(!a)return 3;
        int found=0;
        for(CFIndex i=0;i<CFArrayGetCount(a);i++) {
            CFDictionaryRef d=CFArrayGetValueAtIndex(a,i);
            CFNumberRef p=CFDictionaryGetValue(d,kCGWindowOwnerPID), l=CFDictionaryGetValue(d,kCGWindowLayer);
            int owner=0,layer=-1;CGRect bounds=CGRectZero;
            if(p)CFNumberGetValue(p,kCFNumberIntType,&owner);
            if(l)CFNumberGetValue(l,kCFNumberIntType,&layer);
            CFDictionaryRef b=CFDictionaryGetValue(d,kCGWindowBounds);
            if(b)CGRectMakeWithDictionaryRepresentation(b,&bounds);
            if(owner==pid&&layer==0&&bounds.size.width>=200&&bounds.size.height>=150){found=1;break;}
        }
        CFRelease(a);
        if(found){puts("window");return 0;}
        struct timespec wait={0,50000000};nanosleep(&wait,NULL);
    }
    return 4;
}

File: benchmarks/opt3-gui/window.m
// Diagnostic CI account only. No Accessibility privileges or global app matching.
#import <AppKit/AppKit.h>
#import <CoreGraphics/CoreGraphics.h>
#include <errno.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
int main(int argc,char **argv) {
 @autoreleasepool {
  if(argc!=3)return 2;
  char *end;errno=0;long value=strtol(argv[2],&end,10);
  if(errno||*end||value<1||value>INT_MAX)return 2;
  NSRunningApplication *app=[NSRunningApplication runningApplicationWithProcessIdentifier:(pid_t)value];
  if(!strcmp(argv[1],"hide")||!strcmp(argv[1],"show")||!strcmp(argv[1],"quit")) {
   BOOL ok=NO;
   if(app)ok=!strcmp(argv[1],"hide")?[app hide]:!strcmp(argv[1],"show")?[app unhide]:[app terminate];
   puts(ok?"{\"accepted\":true}":"{\"accepted\":false}");return 0;
  }
  if(strcmp(argv[1],"windows"))return 2;
  CFArrayRef a=CGWindowListCopyWindowInfo(kCGWindowListOptionAll,kCGNullWindowID);
  if(!a)return 3;
  int ordinary=0,visible=0;
  for(CFIndex i=0;i<CFArrayGetCount(a);i++) {
   CFDictionaryRef d=CFArrayGetValueAtIndex(a,i);
   int owner=0,layer=-1;CGRect rect=CGRectZero;
   CFNumberRef p=CFDictionaryGetValue(d,kCGWindowOwnerPID),l=CFDictionaryGetValue(d,kCGWindowLayer);
   if(p)CFNumberGetValue(p,kCFNumberIntType,&owner);
   if(l)CFNumberGetValue(l,kCFNumberIntType,&layer);
   CFDictionaryRef b=CFDictionaryGetValue(d,kCGWindowBounds);
   if(b)CGRectMakeWithDictionaryRepresentation(b,&rect);
   if(owner==value&&layer==0&&rect.size.width>=200&&rect.size.height>=150){
    ordinary++;CFBooleanRef onscreen=CFDictionaryGetValue(d,kCGWindowIsOnscreen);
    if(onscreen&&CFBooleanGetValue(onscreen))visible++;
   }
  }
  CFRelease(a);
  printf("{\"ordinary\":%d,\"onscreen\":%d,\"hidden\":%s}\n",ordinary,visible,app&&app.hidden?"true":"false");
  return 0;
 }
}
